File: src/aws/context.rs
use anyhow::{Result, anyhow};
use aws_config::{BehaviorVersion, SdkConfig};
use once_cell::sync::OnceCell;
use tokio::runtime::Handle;

use crate::aws::{asm::AsmClient, ec2::Ec2Client, imds::ImdsClient, s3::S3Client, ssm::SsmClient};

#[derive(Debug)]
pub struct AwsCtx {
    rt: Handle,
    config: OnceCell<SdkConfig>,
    asm: OnceCell<AsmClient>,
    ec2: OnceCell<Ec2Client>,
    imds: OnceCell<ImdsClient>,
    s3: OnceCell<S3Client>,
    ssm: OnceCell<SsmClient>,
}

impl AwsCtx {
    pub fn new(rt: Handle) -> Result<Self> {
        Ok(Self {
            rt,
            config: OnceCell::new(),
            asm: OnceCell::new(),
            ec2: OnceCell::new(),
            imds: OnceCell::new(),
            s3: OnceCell::new(),
            ssm: OnceCell::new(),
        })
    }

    pub fn asm(&self) -> Result<&AsmClient> {
        let config = self.config()?;
        let client = aws_sdk_secretsmanager::Client::new(config);
        self.asm
            .get_or_try_init(|| Ok(AsmClient::new(self.rt.clone(), client)))
    }

    pub fn ec2(&self) -> Result<&Ec2Client> {
        let config = self.config()?;
        let client = aws_sdk_ec2::Client::new(config);
        self.ec2
            .get_or_try_init(|| Ok(Ec2Client::new(self.rt.clone(), client)))
    }

    pub fn imds(&self) -> Result<&ImdsClient> {
        let client = aws_config::imds::Client::builder().build();
        self.imds
            .get_or_try_init(|| Ok(ImdsClient::new(self.rt.clone(), client)))
    }

    pub fn s3(&self) -> Result<&S3Client> {
        let config = self.config()?;
        let client = aws_sdk_s3::Client::new(config);
        self.s3
            .get_or_try_init(|| Ok(S3Client::new(self.rt.clone(), client)))
    }

    pub fn ssm(&self) -> Result<&SsmClient> {
        let config = self.config()?;
        let client = aws_sdk_ssm::Client::new(config);
        self.ssm
            .get_or_try_init(|| Ok(SsmClient::new(self.rt.clone(), client)))
    }

    fn config(&self) -> Result<&SdkConfig> {
        self.config.get_or_try_init(|| {
            let config = self.rt.block_on(async {
                let config = aws_config::defaults(BehaviorVersion::v2025_08_07())
                    .load()
                    .await;

                let sts = aws_sdk_sts::Client::new(&config);
                sts.get_caller_identity().send().await.map_err(|e| {
                    anyhow!("user data config requires an IAM instance profile: {}", e)
                })?;

                Ok::<_, anyhow::Error>(config)
            })?;
            Ok(config)
        })
    }
}

File: src/aws/context.zig
//! AWS context providing lazily-initialized clients for AWS services.

const std = @import("std");
const Allocator = std.mem.Allocator;

const aws = @import("aws");

const Ec2Client = @import("ec2.zig").Ec2Client;
const S3Client = @import("s3.zig").S3Client;
const SecretsManagerClient = @import("asm.zig").SecretsManagerClient;
const SsmClient = @import("ssm.zig").SsmClient;

const scoped_log = std.log.scoped(.aws_context);

pub const AwsContext = struct {
    allocator: Allocator,

    imds: aws.ImdsClient,
    region: ?[]const u8 = null,
    credentials_verified: bool = false,
    ec2: ?Ec2Client = null,
    s3: ?S3Client = null,
    ssm: ?SsmClient = null,
    secrets_manager: ?SecretsManagerClient = null,

    const Self = @This();

    pub const Error = error{
        /// No IAM instance profile attached to the instance.
        NoInstanceProfile,
    };

    /// Initialize the AWS context with an IMDS client.
    pub fn init(allocator: Allocator) !Self {
        const imds = try aws.ImdsClient.init(allocator, .{});

        return Self{
            .allocator = allocator,
            .imds = imds,
        };
    }

    pub fn deinit(self: *Self) void {
        if (self.ec2) |*client| {
            client.deinit();
        }
        if (self.s3) |*client| {
            client.deinit();
        }
        if (self.ssm) |*client| {
            client.deinit();
        }
        if (self.secrets_manager) |*client| {
            client.deinit();
        }
        if (self.region) |region| self.allocator.free(region);
        self.imds.deinit();
    }

    /// Get the IMDS client.
    pub fn getImds(self: *Self) *aws.ImdsClient {
        return &self.imds;
    }

    /// Get or initialize the S3 client.
    /// Returns error if no IAM instance profile is attached.
    pub fn getS3(self: *Self) !*S3Client {
        try self.verifyCredentials();
        if (self.s3 == null) {
            self.s3 = try S3Client.init(self.allocator, self.region.?);
        }
        return &self.s3.?;
    }

    /// Get or initialize the SSM client.
    /// Returns error if no IAM instance profile is attached.
    pub fn getSsm(self: *Self) !*SsmClient {
        try self.verifyCredentials();
        if (self.ssm == null) {
            self.ssm = try SsmClient.init(self.allocator, self.region.?);
        }
        return &self.ssm.?;
    }

    /// Get or initialize the Secrets Manager client.
    /// Returns error if no IAM instance profile is attached.
    pub fn getSecretsManager(self: *Self) !*SecretsManagerClient {
        try self.verifyCredentials();
        if (self.secrets_manager == null) {
            self.secrets_manager = try SecretsManagerClient.init(
                self.allocator,
                self.region.?,
            );
        }
        return &self.secrets_manager.?;
    }

    /// Get or initialize the EC2 client.
    /// Returns error if no IAM instance profile is attached.
    pub fn getEc2(self: *Self) !*Ec2Client {
        try self.verifyCredentials();
        if (self.ec2 == null) {
            self.ec2 = try Ec2Client.init(self.allocator, self.region.?);
        }
        return &self.ec2.?;
    }

    fn verifyCredentials(self: *Self) !void {
        if (self.credentials_verified) return;
        try self.resolveRegion();
        var creds = self.imds.getIamCredentials(.{}) catch |err| {
            scoped_log.err(
                "user data config requires an IAM instance profile: {s}",
                .{@errorName(err)},
            );
            return Error.NoInstanceProfile;
        };
        creds.deinit();
        self.credentials_verified = true;
    }

    fn resolveRegion(self: *Self) !void {
        if (self.region != null) return;
        const region = self.imds.getMetadata(
            "/latest/meta-data/placement/region",
            .{},
        ) catch |err| {
            scoped_log.err("failed to get region from IMDS: {s}", .{@errorName(err)});
            return err;
        };
        scoped_log.info("AWS region: {s}", .{region});
        self.region = region;
    }
};
